### PR TITLE
Blog Post: Fetch API for building servers

### PR DESCRIPTION
**Pull request by ardatan**
Title suggestions:
- From @charlypoly 
    ”Running JavaScript servers everywhere with Fetch API”
    ”The Future of JavaScript servers is platform-agnostic: Fetch API”
    ”A proposal for a the future of JavaScript servers”
   
- From @n1ru4l 
    Running JavaScript HTTP servers anywhere with the Fetch API
    
- From @Urigo 
    JavaScript runs everywhere, so should your servers - here is how
    
From @n1ru4l 
Maybe just inline the first yoga example and add these others as a “Bonus” section at the end of the blog post? Dunno whether it is a good idea to link to a lot of code here.

### CODE REVIEW

File: pages/blog/fetch-for-servers.mdx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+At the beginning of the year, we’ve launched GraphQL Yoga 2.0 - a server framework for GraphQL APIs.
+
+While planning version 2.0 of Yoga, we were thinking about all the things that changed in the ecosystem and what developers using JavaScript expect from their server frameworks now and in the future.
+
+One of the most powerful trend in the JS ecosystem was the proliferation of new environments and platforms that can run JS(Lambdas, Cloudflare Workers, Deno, Bun etc.). So we set up to build a single GraphQL server framework that could run on any of these platforms.
+
+What we’ve found in the process was fascinating and we believe would change how any JS HTTP frameworks are being built, without any relation to GraphQL
+
+While Node.js is the most popular environment, multiple platforms can run JavaScript with different APIs. Usually, they have their own way of creating servers, but we saw they share a web standard that most people usually use for communicating with the backends from the client side, not the other way around. That’s WHATWG Fetch API that uses `Request`, `Response`, `ReadableStream`, etc to manage this on the server side, too.
+
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:26]: I'm not sure what you meant here so I suggested some changes by having some assumptions:

```suggestion
While Node.js is the most popular environment, multiple platforms can run JavaScript, usually, having their own way of creating servers, and using different APIs.

On the other hand, the JavaScript client-side has mostly migrated over the years to a common set of standards (`fetch`), independently of the underlying platform.

This is where we realized that WHATWG Fetch API that uses `Request`, `Response`, and `ReadableStream` could also be used on the server side to provide a unified API to run JavaScript HTTP servers everywhere.

```

File: pages/blog/fetch-for-servers.mdx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+## Why Fetch API?
+
+When you send a request from the client, you basically use `fetch(...requestArgs)` that uses `Request` object under the hood that contains all the details(headers, method, etc) and the data stream you need for that communication, then you take `Response` object that contains the connection stream and it processes it as you want with `.json`, `.arrayBuffer`, `.formData` or you can just access the stream itself by `.body` as `ReadableStream`
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:29]:
```suggestion
When you send a request from the client, `fetch(...requestArgs)`  uses `Request` object under the hood that contains all the details(headers, method, etc) and the data stream you need for that communication. Then you take the `Response` object that contains the connection stream, and process it as you want with `.json()`, `.arrayBuffer()`, `.formData()` or just access the stream itself by `.body()` as `ReadableStream`
```

File: pages/blog/fetch-for-servers.mdx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+On the server side, you can take `Request` object as a server and then process it with the same methods without dealing with the internals of your platform.
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:31]:
```suggestion
On the server side, you can take `Request` object and process it with the same methods without dealing with the internals of your platform.
```

File: pages/blog/fetch-for-servers.mdx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+Streaming responses like SSE, we just use `ReadableStream` and for Multipart requests (e.g. file uploads), we just take `FormData` which is exactly the same passed from the browser or any other client uses Fetch API.
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:33]:
```suggestion
Streaming responses like SSE uses `ReadableStream`, and for Multipart requests (e.g. file uploads)uses  `FormData` which is exactly the same forwarded from the browser or any other client using Fetch API.
```

File: pages/blog/fetch-for-servers.mdx [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+SSE; [https://github.com/dotansimha/graphql-yoga/blob/master/packages/common/src/plugins/resultProcessor/push.ts#L42](https://github.com/dotansimha/graphql-yoga/blob/master/packages/common/src/plugins/resultProcessor/push.ts#L42)
+
+## What about Node.js?
+
+Currently, in the older LTS versions of Node.js, we don’t have an implementation of the Fetch API builtin. Furthermore, Node.js doesn’t use Web standard streams and the Fetch API in its `http` and `https` modules. That’s why we created the `@whatwg-node/fetch` package (previously known as `cross-undici-fetch` ) that fills in the gaps of different `fetch` implementations in all the LTS Node.js versions. Under the hood, `@whatwg-node/fetch` utilizes `undici` if available or otherwise falls back to using `node-fetch`, which you are probably already familiar with. In case `@whatwg-node/fetch` is imported in an environment that already has Fetch API built in like Cloudflare Workers, no ponyfills are added to your built application bundle.
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:41]:
```suggestion
Currently, in the older LTS versions of Node.js, we don’t have an implementation of the Fetch API built in. Furthermore, Node.js doesn’t use Web standard streams and the Fetch API in its `http` and `https` modules. That’s why we created the `@whatwg-node/fetch` package (previously known as `cross-undici-fetch` ) that fills in the gaps of different `fetch` implementations in all the LTS Node.js versions. Under the hood, `@whatwg-node/fetch` utilizes `undici` if available or otherwise falls back to using `node-fetch`, which you are probably already familiar with. In case `@whatwg-node/fetch` is imported in an environment that already has Fetch API built in like Cloudflare Workers, no ponyfills are added to your built application bundle.
```

File: pages/blog/fetch-for-servers.mdx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+- Ponyfill vs Polyfill
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:43]:
```suggestion
**Ponyfill vs Polyfill**
```

File: pages/blog/fetch-for-servers.mdx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  Polyfill patches the native parts of an environment while ponyfill just exports the “patched” stuff without touching the environment’s internals. We prefer pony filling because it prevents us from breaking other libraries and environmental functionalities.
+
+## Is it possible to have a library that creates a cross-platform server?
+
+When rebuilding GraphQL Yoga from scratch, this was one of the most important features we wanted to implement. We wanted to create a GraphQL server library that can be integrated with different Node.js server frameworks, and other JS environments like CF Workers and Deno with a few additional lines of code.
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:48]:
```suggestion
When rebuilding GraphQL Yoga from scratch, cross-platform support was one of the most important features we wanted to implement. We wanted to create a GraphQL server library that can be integrated with different Node.js server frameworks and other JS environments like CF Workers and Deno with a few additional lines of code.
```

File: pages/blog/fetch-for-servers.mdx [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+Finally how small the code is when we want to process the request and the response objects;
+
+[https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins](https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins)
+
+> There is literally nothing platform-specific in GraphQL Yoga, and this allows us to focus on creating a good GraphQL Server implementation for the entire GraphQL JS ecosystem.
+
+## Node.js Server Frameworks, Routers & Middlewares
+
+Node.js was there in the JS community for years before the web standards were not as complete as today. We think Node.js inspired a lot of ideas on how server-side stuff can be easily managed with JavaScript by implementing all the Stream and HTTP APIs. But now we have Fetch, Web Streams, and so on as a web standard in the JavaScript ecosystem.
```
Comment (charlypoly) [inline pages/blog/fetch-for-servers.mdx:106]:
```suggestion
Node.js solved issues years before the web standards could. However, we think that many server-side ideas inspired by Node.js can now easily be managed with JavaScript with Fetch, Web Streams, and so on as a web standard in the JavaScript ecosystem.
```

**Review by charlypoly — CHANGES_REQUESTED** [review]
That's an awesome inspiring article! 🚀 

I've left some phrasing suggestions

File: pages/blog/fetch-for-servers.mdx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  description: 'The new way of building servers in JavaScript',
+  image:
+    'https://user-images.githubusercontent.com/727224/104328926-56b0a700-54ba-11eb-9ad3-f6d2b971d55d.jpg',
+  thumbnail:
+    'https://user-images.githubusercontent.com/727224/104328777-29fc8f80-54ba-11eb-87d0-9982af1d8281.jpg'
+}
+
+export default withArticle(meta)
+
+At the beginning of the year, we’ve launched GraphQL Yoga 2.0 - a server framework for GraphQL APIs.
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:17]: link to the blog post about graphql yoga

File: pages/blog/fetch-for-servers.mdx [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  date: '2022-07-21',
+  description: 'The new way of building servers in JavaScript',
+  image:
+    'https://user-images.githubusercontent.com/727224/104328926-56b0a700-54ba-11eb-9ad3-f6d2b971d55d.jpg',
+  thumbnail:
+    'https://user-images.githubusercontent.com/727224/104328777-29fc8f80-54ba-11eb-87d0-9982af1d8281.jpg'
+}
+
+export default withArticle(meta)
+
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:16]:
```suggestion
## TLDR
* We made GraphQL Yoga 2.0 **platform agnostic**, so it is able to run anywhere JavaScript can run (Cloudflare Workers, sdfsdf......) thanks to the new **Fetch API standard**
* We've created <ins>Ponyfills</ins> so it will work the same on older versions of Node that don't support the standard
* We've made a **new library** out of it so that any framework and app could achieve the same
* Help other frameworks in the ecosystem to migrate to this new library and standard
```

File: pages/blog/fetch-for-servers.mdx [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+const myServer = createServerAdapter({
+  handleRequest: router.handle
+})
+
+// Then use it in any environment
+import { createServer } from 'http'
+
+const httpServer = createServer(myServer)
+httpServer.listen(4000)
+```
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:175]:
```suggestion
What NodeJS frameworks are you using today?
Maybe it is worth while to open a new issue on their repo and see if we could all help them to become platform agnostic, while still supporting older versions of Node, thanks to this new library!
```

**Review by Urigo — CHANGES_REQUESTED** [review]

File: pages/blog/fetch-for-servers.mdx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+export const meta = {
+  title: 'JavaScript runs everywhere, so should your servers - here is how',
+  tags: ['fetch', 'whatwg', 'node', 'graphql', 'graphql-yoga'],
+  authors: ['arda'],
+  date: '2022-07-21',
+  description: 'The new way of building servers in JavaScript',
+  image:
+    'https://user-images.githubusercontent.com/727224/104328926-56b0a700-54ba-11eb-9ad3-f6d2b971d55d.jpg',
+  thumbnail:
+    'https://user-images.githubusercontent.com/727224/104328777-29fc8f80-54ba-11eb-87d0-9982af1d8281.jpg'
```
Comment (n1ru4l) [inline pages/blog/fetch-for-servers.mdx:12]: We still need proper social images here.

File: pages/blog/fetch-for-servers.mdx (hunk 9 above)
Comment (charlypoly) [reply pages/blog/fetch-for-servers.mdx:21]:
```suggestion
At the beginning of the year, [we launched GraphQL Yoga 2.0](https://www.the-guild.dev/blog/announcing-graphql-yoga-v2) - a server framework for GraphQL APIs.
```

**Review by charlypoly — APPROVED** [review]

**Comment by charlypoly** [comment]
@ardatan, I fixed the Vercel build

Just a few last comments from @Urigo and I, a proper cover illustration, and we're good to go! 🚀

**Comment by charlypoly** [comment]
@ardatan, here is some proposal of illustrations (title of the article to be confirmed):

caching
The Anatomy of a GraphQL Request-1000x400 (3)

**Comment by erikash** [comment]
Loved every bit of it!
I've been using my own `Request` and `Response` classes with adapters for different libraries and this take is much much better!

**Comment by charlypoly** [comment]
@ardatan I've update the PR with proper images

**Comment by ardatan** [comment]
@charlypoly Thank you, looks way better <3

File: pages/blog/fetch-for-servers.mdx [hunk 13]
```diff
@@ -0,0 +1,182 @@
+import withArticle from '../../ui/blog/article'
+
+export const meta = {
+  title: 'JavaScript runs everywhere, so should your servers - here is how',
```
Comment (n1ru4l) [inline pages/blog/fetch-for-servers.mdx:4]:
```suggestion
  title: 'JavaScript runs everywhere, so should your HTTP server - here is how',
```

I think this is more specific

File: pages/blog/fetch-for-servers.mdx [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+As we already mentioned in the “Why Fetch API?” part, the server library itself doesn’t need to care about the platform’s connection-specific details like Node’s `IncomingMessage` and `ServerResponse` or Next.js’s `NextApi.Request` objects. You are now able to focus on your server implementation details by consuming a “universal standard” [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) and returning “another standard” [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response)
+
+As we realized, how well this works out for our users, we decided that we need to bring this to the next level and extract that logic into a standalone library called `@whatwg-node/server` . You simply provide your request handler that has a single [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) parameter and expects you to return a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) instance. The generated request handler instance can be integrated with regular Node HTTP servers, Fastify, Koa, Deno, CF Workers, Next.js, etc with a few lines of code.
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:92]:
```suggestion
As we realized how well this works out for our users, we decided that we need to bring this to the next level and extract that logic into a standalone library called [`@whatwg-node/server` ](https://github.com/ardatan/whatwg-node). 

You simply provide your request handler that has a single [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) parameter and expects you to return a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) instance. The generated request handler instance can be integrated with regular Node HTTP servers, Fastify, Koa, Deno, CF Workers, Next.js, etc with a few lines of code.
```

File: pages/blog/fetch-for-servers.mdx [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+
+Finally how small the code is when we want to process the request and the response objects;
+
+[https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins](https://github.com/dotansimha/graphql-yoga/tree/no-more-node/packages/graphql-yoga/src/plugins)
+
+> There is literally nothing platform-specific in GraphQL Yoga, and this allows us to focus on creating a good GraphQL Server implementation for the entire GraphQL JS ecosystem.
+
+## Node.js Server Frameworks, Routers & Middlewares
+
+Node.js solved issues years before the web standards could. However, we think that many server-side ideas inspired by Node.js can now easily be managed with JavaScript with Fetch, Web Streams, and so on as a web standard in the JavaScript ecosystem.
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:142]:
```suggestion
Node.js solved issues years before the web standards could. However, we think that many server-side ideas inspired by Node.js can now easily be managed with JavaScript with Fetch, Web Streams and other web standards in the JavaScript ecosystem.
```

File: pages/blog/fetch-for-servers.mdx [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+
+There are many mature libraries such as Fastify, Koa, Express, and Hapi that are implemented only for Node.js which Fetch API doesn’t have in its ecosystem for now. The experience with those libraries in the current era of Node.js taught us a lot about how the server can be designed but maybe it is time to reduce the environment-specific APIs in the JS ecosystem.
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:144]:
```suggestion
There are many mature libraries such as Fastify, Koa, Express, and Hapi that are implemented only for Node.js without using the Fetch API Standard. The experience with those libraries in the current era of Node.js taught us a lot about how the server can be designed but maybe it is time to reduce the environment-specific APIs in the JS ecosystem.
```

File: pages/blog/fetch-for-servers.mdx [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+
+The major reason for using a server framework is usually “Routing” then “Middlewares” so the question is “Why cannot we just have that with Fetch API?”
+
+You can basically achieve routing like below however it looks a bit unsafe.
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:148]:
```suggestion
You can basically achieve routing like below, however it looks a bit unsafe.
```

File: pages/blog/fetch-for-servers.mdx [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+// Then use it in any environment
+import { createServer } from 'http'
+
+const httpServer = createServer(myServer)
+httpServer.listen(4000)
+```
+
+## What NodeJS frameworks are you using today?
+
+Maybe it is worth while to open a new issue on their repo and see if we could all help them to become platform agnostic, while still supporting older versions of Node, thanks to this new library!
```
Comment (Urigo) [inline pages/blog/fetch-for-servers.mdx:211]:
```suggestion
Maybe it is worth while to open a new issue on their repo and see if we could all help them to become platform agnostic, while still supporting older versions of Node, thanks to this new library.

Please try it out and give us feedback [on the repo](https://github.com/ardatan/whatwg-node)!
```

File: pages/blog/fetch-for-servers.mdx [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+  description: 'The new way of building servers in JavaScript',
+  image: '/blog-assets/fetch-for-servers/cover.png',
+  thumbnail: '/blog-assets/fetch-for-servers/thumbnail.png'
+}
+
+export default withArticle(meta)
+
+## TLDR
+
+- We made GraphQL Yoga 2.0 **platform agnostic**, so it is able to run anywhere JavaScript can run (Cloudflare Workers, sdfsdf......) thanks to the new **Fetch API standard**
```
[the remark this answers is not in the document]
Comment (ardatan) [reply pages/blog/fetch-for-servers.mdx:17]: Oh my bad. I forgot it while writing the draft 🤦

File: pages/blog/fetch-for-servers.mdx [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+## Why Fetch API?
+
+When you send a request from the client, `fetch(...requestArgs)` uses [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object under the hood that contains all the details(headers, method, etc) and the data stream you need for that communication. Then you take the [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object that contains the connection stream, and process it as you want with `.json()`, `.arrayBuffer()`, `.formData()` or just access the stream itself by `.body()` as `ReadableStream`
+
+On the server side, you can take [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object and process it with the same methods without dealing with the internals of your platform.
+
+Streaming responses like SSE uses `ReadableStream`, and for Multipart requests (e.g. file uploads)uses `FormData` which is exactly the same forwarded from the browser or any other client using Fetch API.
```
[the remark this answers is not in the document]
Comment (ardatan) [reply pages/blog/fetch-for-servers.mdx:41]: I mean you can get the exactly same `FormData` on the server side with `request.formData`.

File: pages/blog/fetch-for-servers.mdx [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+
+export default withArticle(meta)
+
+## TLDR
+
+- We made GraphQL Yoga 2.0 **platform agnostic**, so it is able to run anywhere JavaScript can run (Cloudflare Workers, sdfsdf......) thanks to the new **Fetch API standard**
+- We've created <ins>Ponyfills</ins> so it will work the same on older versions of Node that don't support the standard
+- We've made a **new library** out of it so that any framework and app could achieve the same
+- Help other frameworks in the ecosystem to migrate to this new library and standard
+  At the beginning of the year, [we launched GraphQL Yoga 2.0](https://www.the-guild.dev/blog/announcing-graphql-yoga-v2) - a server framework for GraphQL APIs.
```
[the remark this answers is not in the document]
Comment (ardatan) [reply pages/blog/fetch-for-servers.mdx:21]: I actually do some introduction below but not sure what do you mean by a header actually?

File: pages/blog/fetch-for-servers.mdx (hunk 13 above)
Comment (ardatan) [reply pages/blog/fetch-for-servers.mdx:4]: Hmm we need a new image then @Urigo @charlypoly What do you think?

Comment (charlypoly) [reply pages/blog/fetch-for-servers.mdx:4]: I guess we can, let me know if you need a new set of images

Comment (Urigo) [reply pages/blog/fetch-for-servers.mdx:4]: I prefer the existing title.....

**Review by Urigo — APPROVED** [review]
